### PR TITLE
feat: Add transaction message filter in GraphQL

### PR DESCRIPTION
**Pull request by jinoosss**
## Description

I added the data that is fetched by the query.
Then I decoded and processed the data in [std.Tx](https://github.com/gnolang/gno/blob/676026530df691039b560fa37b39d9aa05a58f46/tm2/pkg/std/tx.go#L15-L20) and added filterable parameters.
- Add transaction message data 
- Implement the filtering functions of the transaction filter model
- Separate the GraphQL schema 

#### Querying transaction messages 
```
{
  transactions(filter: {}) {
    hash
    # messages 
    messages {
      route
      typeUrl
      value {
        # Data in bank.MsgSend to query  
        ... on BankMsgSend {
          from_address
          to_address
          amount
        }

        # Data in vm.m_call to query  
        ... on MsgCall {
          send
          pkg_path
          func
          caller
          args
        }

        # Data in vm.m_addpkg to query  
        ... on MsgAddPackage {
          creator
          package {
            Name
            Path
            Files {
              Name
              Body
            }
          }
          deposit
        }

        # Data in vm.m_run to query  
        ... on MsgRun {
          caller
          send
          package {
            Name
            Path
            Files {
              Name
              Body
            }
          }
        }
      }
    }
  }
}
```

#### Filter transactio messages 
```
{
  transactions(
    filter: {
      message: {
        route: bank 		# bank / vm
        type_url:  send	# send / exec / add_pacakge / run

        # Parameters for messages with route as `bank`  
        bank_param: {
          send: {
            # ...
          }
        }

        # Parameters for messages with route as `vm`
        vm_param: {
          m_call: {
            # ...
          }
          m_addpkg: {
            # ...
          }
          m_run: {
            # ...
          }
        }
      }
    }
  ) {
    hash
  }
}
```
image

### CODE REVIEW

File: serve/graph/model/models_gen.go [hunk 1]
```diff
@@ -3,19 +3,113 @@
...
+	Caller *string `json:"caller,omitempty"`
+	// Filter by `send`.
+	Send *string `json:"send,omitempty"`
+	// Filter by `pkg_path`.
+	PkgPath *string `json:"pkg_path,omitempty"`
+	// Filter by `func`.
+	Func *string `json:"func,omitempty"`
+	// Filter by `args`, Arguments are checked in the order of the argument array, and arguments that are not checked are left blank.
+	Args []string `json:"args,omitempty"`
+}
```
Comment (ajnavarro) [inline serve/graph/model/models_gen.go:95]: The same as previous comments. These Go docs will be shown to the user as API documentation. Would be great if you could give more info.

File: serve/graph/model/transaction.go [hunk 2]
```diff
@@ -4,41 +4,133 @@ import (
...
 
+	"github.com/gnolang/gno/gno.land/pkg/sdk/vm"
+	"github.com/gnolang/gno/tm2/pkg/amino"
 	"github.com/gnolang/gno/tm2/pkg/bft/types"
+	"github.com/gnolang/gno/tm2/pkg/sdk/bank"
+	"github.com/gnolang/gno/tm2/pkg/std"
 )
 
 type Transaction struct {
-	t *types.TxResult
```
Comment (ajnavarro) [inline serve/graph/model/transaction.go:11]: Why is this change needed?

File: serve/graph/model/utils.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package model
```
Comment (ajnavarro) [inline serve/graph/model/utils.go:1]: No utils files, please. They end up being kitchen-sinks, where everything can be.

**Review by ajnavarro — CHANGES_REQUESTED** [review]
The feature added seems ok, but it is really difficult to review because we have on the same PR a refactoring and a feature.

I would suggest keeping this PR as a feature implementation and avoiding refactors on the code.

File: serve/graph/model/models_gen.go [hunk 4]
```diff
@@ -3,19 +3,113 @@
...
+
+// Input parameters required when the message type is `send`.
+type BankMsgSendInput struct {
+	// Filter by `from_address`.
+	FromAddress *string `json:"from_address,omitempty"`
+	// Filter by `to_address`.
+	ToAddress *string `json:"to_address,omitempty"`
+	// Filter by `amount`.
+	Amount *string `json:"amount,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/model/models_gen.go:31]: I forgot that this code is generated from the schema, so the schema is the file that should be changed/improved documentation-wise.

**Comment by jinoosss** [comment]
@ajnavarro, 
Thank you for your thorough review 🙏 

Your comments have given me some direction. 
It seems that the model is sufficient as a data passer, and I've changed it so that the resolver can handle the functionality.

File: serve/graph/model/utils.go (hunk 3 above)
Comment (jinoosss) [reply serve/graph/model/utils.go]: I removed the utils file and moved the deref function to resolver.

File: gqlgen.yml [hunk 5]
```diff
@@ -1,11 +1,12 @@
 # Where are all the schema files located? globs are supported eg  src/**/*.graphql
 schema:
-  - serve/graph/*.graphql
+  - serve/graph/schema/*.graphql
+  - serve/graph/schema/**/*.graphql
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply gqlgen.yml:4]: As I had more filters and schema types, I realized it would be nice to separate them.

So I added a subdirectory to get the schema's subdirectories.

File: gqlgen.yml [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 
 # Where should the generated server code go?
 exec:
-  filename: serve/graph/generated.go
-  package: graph
+  filename: serve/graph/generated/generated.go
+  package: generated
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply gqlgen.yml:9]: `setup.go` needs to import `resolver` and `generated`, and `resolver` imports `generated`. 
I separated setup and generated into separate packages because putting them at the same depth would cause import errors.

File: serve/graph/model/block_filter.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	return uint64(Deref(filter.ToHeight))
+}
+
+func (filter *BlockFilter) GetFromTime() time.Time {
+	return Deref(filter.FromTime)
+}
+
+func (filter *BlockFilter) GetToTime() time.Time {
+	return Deref(filter.ToTime)
+}
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/model/block_filter.go]: I changed it.

File: serve/graph/model/block_filter.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+func (filter *BlockFilter) FilterBy(block *types.Block) bool {
+	return filter.filterByBlockTime(block.Time)
+}
+
+func (filter *BlockFilter) filterByBlockTime(blockTime time.Time) bool {
+	fromTime := filter.GetFromTime()
+	toTIme := filter.GetToTime()
+
+	return (blockTime.After(fromTime) || blockTime.Equal(fromTime)) && (toTIme.IsZero() || blockTime.Before(toTIme))
+}
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/model/block_filter.go]: I changed it.

File: serve/graph/model/models_gen.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/model/models_gen.go]: I added comments to the graphql schema.

File: serve/graph/model/models_gen.go [hunk 9]
```diff
@@ -3,19 +3,113 @@
...
+
+// Input parameters required when the message type is `add_package`.
+type MsgAddPackageInput struct {
+	// Filter by `creator`.
+	Creator *string `json:"creator,omitempty"`
+	// Filter by `package`.
+	Package *MemPackageInput `json:"package,omitempty"`
+	// Filter by `deposit`.
+	Deposit *string `json:"deposit,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/model/models_gen.go]: I added comments to the graphql schema.

File: serve/graph/model/models_gen.go (hunk 1 above)
Comment (jinoosss) [reply serve/graph/model/models_gen.go]: I added comments to the graphql schema.

File: gqlgen.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajnavarro) [reply gqlgen.yml:9]: If `resolver.go` and `generated.go` are on the same package as before `graph` then a cyclic import is not possible. `resolver.go` and `generated.go` does not need to import anything because they are on the same package, and `setup.go` can import graph, having methods from both without any error.

File: serve/graph/model/transaction.go (hunk 2 above)
Comment (jinoosss) [reply serve/graph/model/transaction.go:11]: This doesn't need to be changed, 
but I thought there might be some confusion when implementing the function, so I made a small modification.

File: gqlgen.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jinoosss) [reply gqlgen.yml:9]: You're right. 
If the three are managed in the same package, you won't get the cycle import error. 
Do you think it's good to have `resolver`, `genertated`, and `setup` managed in the same package?

[the remark this answers is not in the document]
Comment (ajnavarro) [reply gqlgen.yml:9]: yes, as it was before if I'm remembering correctly, right?

[the remark this answers is not in the document]
Comment (jinoosss) [reply gqlgen.yml:9]: Yes, that's correct.
Then I'll change it back to managing them in one package.

I think it's better to keep them separate, but I'll think about it more.

File: serve/graph/subscription.resolvers.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/gnolang/tx-indexer/serve/graph/model"
+	"github.com/gnolang/tx-indexer/types"
+)
+
+// Transactions is the resolver for the transactions field.
+func (r *subscriptionResolver) Transactions(ctx context.Context, filter model.TransactionFilter) (<-chan *model.Transaction, error) {
+	return handleChannel(ctx, r.manager, func(nb *types.NewBlock, c chan<- *model.Transaction) {
+		for _, tx := range nb.Results {
+			transactionResolver := NewTransactionResolver(model.NewTransaction(tx))
```
Comment (ajnavarro) [inline serve/graph/subscription.resolvers.go:18]: This is memory inefficient hitting badly to the Go GC. You are creating a new object per result obtained from the storage, only needed to apply a filter. WDYT about removing the TransactionResolver struct and the BlockResolver struct and just leaving filter methods? Example:

```go

func FilterTransactionBy(tx *model.Transaction, f model.TransactionFilter) bool {...}

```

File: serve/graph/model/transaction.go [hunk 11]
```diff
@@ -2,43 +2,228 @@ package model
...
+	var stdTx std.Tx
+
+	if err := amino.Unmarshal(txResult.Tx, &stdTx); err != nil {
+		return nil
+	}
+
+	messages := make([]*TransactionMessage, 0)
+	for _, message := range stdTx.GetMsgs() {
+		messages = append(messages, NewTransactionMessage(message))
+	}
```
Comment (ajnavarro) [inline serve/graph/model/transaction.go:31]: Could we make this operation lazy? we can only apply the unmarshalling when needed, that is, when calling Messages() method. We can use `sync.Once` for that. Queries not reading messages will be much faster.

File: serve/graph/model/transaction.go [hunk 12]
```diff
@@ -2,43 +2,228 @@ package model
...
+
+	switch message.Route() {
+	case bank.RouterKey:
+		if message.Type() == MessageTypeSend.String() {
+			contentMessage = &TransactionMessage{
+				Route:   MessageRouteBank,
+				TypeURL: MessageTypeSend,
+				Value:   makeBankMsgSend(message),
+			}
+		}
```
Comment (ajnavarro) [inline serve/graph/model/transaction.go:96]: What if the message type is not TypeSend? Do we lose the message?

Maybe we need a message type unknown for these cases, just in case we receive something that the indexer does not understand yet, to avoid the case that we are hiding information to the user.

The unknown message will only contain the type, route and the rawContent.

**Review by ajnavarro — CHANGES_REQUESTED** [review]
Really good work you did here. Just a few more comments UX related, and performance-related.

Thank you very much!

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+input BankMsgSendInput {
+  """
+  the bech32 address of the fund sender.
+  """
+  from_address: String
+
+  """
+  the bech32 address of the fund receiver.
+  """
+  to_address: String
```
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/schema/filter/transaction_filter.graphql:108]: Also, how is the address range working? are they byte-ordered? Are they lexicographically ordered? `from` is inclusive or exclusive? `to` is inclusive or exclusive?

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+
+  """
+  `TransactionBankMessageInput` represents input parameters required when the message router is `bank`.
+  """
+  bank_param: TransactionBankMessageInput
+
+  """
+  `TransactionVmMessageInput` represents input parameters required when the message router is `vm`.
+  """
+  vm_param: TransactionVmMessageInput
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:83]: I think it's a good idea. 

I thought about making it simpler. 
But it seemed like it would be hard to guarantee the structure of the transaction message.

For example.
Currently, only the `bank` router has messages of type `send`, but it is possible that the `vm` router will also have messages of type `send`. In this case, it would be the same `send` type, but with a different payload.

Of course, this is very unlikely to happen, but I thought we should also consider exceptions.

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 13 above)
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:108]: > How are these values expected? base64? hex? json? Could you add that info tho the param description?

I think it definitely needs more explanation too. 

I'm trying to change it to something like below, but I'm not sure what part I need to explain.
Could you please give me some feedback ?
```
input BankMsgSendInput {
  """
  the bech32 address of the fund sender.
  You can filter by the fund sender address.
  ex) `g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5`
  """
  from_address: String

  """
  the bech32 address of the fund receiver.
  You can filter by the fund receiver address.
  ex) `g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5`
  """
  to_address: String
  
  ...
}
```

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+  """
+  the bech32 address of the fund receiver.
+  """
+  to_address: String
+
+  """
+  the denomination and amount of fund sent ("<amount><denomination>").
+  """
+  amount: String
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:113]: > Also, what do you think about making it possible to filter by a range of amount values? I think that will be much more useful for the end user. Proposal:
> 
> ```
> input AmountFilter {
>     from: int
>     to: int
>     denomination: String
> }
> ```
> 
> We can use in other places, like `deposit` on add package message.

I think that's a really good idea.
Let's add this.

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+  """
+  `MsgAddPackageInput` represents input parameters required when the message type is `add_package`.
+  """
+  m_addpkg: MsgAddPackageInput
+
+  """
+  `MsgRunInput` represents input parameters required when the message type is `run`.
+  """
+  m_run: MsgRunInput
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:133]: Those names are the endpoint contents of the typeUrl when creating the transaction.
ex) vm.m_call, vm.m_addpkg ...

But now it seems better to specify them as `exec`, `add_package`, `run`. Because we already use these values.

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+
+  """
+  Hash from Transaction content in base64 encoding. If this filter is used, any other filter will be ignored.
+  """
+  hash: String
+
+  """
+  Transaction's message to filter Transactions.
+  """
+  message: TransactionMessageInput
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:53]: I felt that the name filter was closer to a feature.

So the higher-level input models, TransactionFilter, BlockFilter, etc. felt like models for functionality, 
while the lower-level input models felt like simple input data models rather than filters.

File: serve/graph/model/transaction.go (hunk 12 above)
Comment (jinoosss) [reply serve/graph/model/transaction.go]: I changed it to show the type and raw data if the message was not generated.

File: serve/graph/model/transaction.go (hunk 11 above)
Comment (jinoosss) [reply serve/graph/model/transaction.go]: I changed it to only decode when accessing stdTx and messages.

File: serve/graph/subscription.resolvers.go (hunk 10 above)
Comment (jinoosss) [reply serve/graph/subscription.resolvers.go]: I removed the resolver and changed it to a function.

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+"""
+`MsgAddPackageInput` represents input parameters required when the message type is `add_package`.
+"""
+input MsgAddPackageInput {
+  """
+  the package deployer.
+  """
+  creator: String
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql]: 'creator' is the address of the person deploying the package.
The form of the address is a bech32-encoded string.
I added a description.

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+
+  """
+  the function name being invoked.
+  """
+  func: String
+
+  """
+  `args` are the arguments passed to the executed function.
+  Arguments are checked in the order of the argument array, and arguments that are not checked are left blank.
+  """
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql]: I realized that seemed wrong to me, too, 
so I fixed the comment.

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+input TransactionMessageInput {
+  """
+  The type of transaction message.
+  """
+  type_url: MessageType
+
+  """
+  The route of transaction message.
+  """
+  route: MessageRoute
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql]: I modified the description.

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 14 above)
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/schema/filter/transaction_filter.graphql:89]: We can even simplify more:

```go
query {
    transactions(filter:{
        // route: vm // you can still filter by vm or bank
        params: {
              from_address: "VAL"
              caller: "CALLER"
        }
    })
}
```

Now, we don't need to care about having different send parameters with different payloads. With this approach, every transaction with a param `from_address!=VAL AND caller!=CALLER` will be filtered, even if we have these same params on different kinds of messages.

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 13 above)
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/schema/filter/transaction_filter.graphql]: Sorry, my bad, I thought that bench32 was a _kind_ of address format, not a serialization format. What you proposed is good, adding an example.

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 16 above)
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/schema/filter/transaction_filter.graphql]: Yeah, LGTM

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+  from: Int
+
+  """
+  The maximum quantity of tokens to check for.
+  """
+  to: Int
+
+  """
+  Filter by token's denomination.
+  If set to an empty string, it will get an empty value.
```
Comment (ajnavarro) [inline serve/graph/schema/filter/transaction_filter.graphql:280]:
```suggestion
  If set to an empty string, denomination won't be considered, so amounts from different currencies will be shown.
```

**Review by ajnavarro — APPROVED** [review]
LGTM apart from a final request:

- [Filter by params](https://github.com/gnolang/tx-indexer/pull/20/files#r1551832344). Let me know what you think about that idea.

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 14 above)
[the remark this answers is not in the document]
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:89]: It certainly seems simple enough.
But I'm worried about confusing the filter parameters.

My concern is twofold 
1. users should be able to recognize exactly what parameters they are filtering. 
For example, amount in BankMsgSend, send in MsgCall, etc., can be missed if they are separated by variable names, and they need to know the description and structure.

2. It is difficult to distinguish conditions between parameters with the same name. 
For example, if you only want to filter messages from MsgCall and MsgPackage by package path, there is no way to filter messages from MsgRun.

Can you share your thoughts ?

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+  """
+  the gno path of the package.
+  """
+  Path: String
+
+  """
+  the associated package gno source.
+  """
+  Files: [MemFileInput]
+}
```
Comment (ajnavarro) [inline serve/graph/schema/filter/transaction_filter.graphql:247]:
```suggestion
input MemPackageInput {
  """
  the name of the package.
  """
  name: String

  """
  the gno path of the package.
  """
  path: String

  """
  the associated package gno source.
  """
  files: [MemFileInput]
}
```

File: serve/graph/schema/filter/transaction_filter.graphql [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+  """
+  the name of the source file.
+  """
+  Name: String
+
+  """
+  the content of the source file.
+  """
+  Body: String
+}
```
Comment (ajnavarro) [inline serve/graph/schema/filter/transaction_filter.graphql:262]:
```suggestion
"""
`MemFileInput` is the metadata information tied to a single gno package / realm file.
"""
input MemFileInput {
  """
  the name of the source file.
  """
  name: String

  """
  the content of the source file.
  """
  body: String
}
```

File: serve/graph/schema/types/transaction.graphql [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+  """
+  the name of the source file.
+  """
+  Name: String!
+
+  """
+  the content of the source file.
+  """
+  Body: String!
+}
```
Comment (ajnavarro) [inline serve/graph/schema/types/transaction.graphql:253]:
```suggestion
"""
`MemFile` is the metadata information tied to a single gno package / realm file
"""
type MemFile {
  """
  the name of the source file.
  """
  name: String!

  """
  the content of the source file.
  """
  body: String!
}
```

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 22 above)
Comment (jinoosss) [reply serve/graph/schema/filter/transaction_filter.graphql:247]: The conventions for MemPackage, MemFile followed gno's json types.
The actual json response comes in Pascalcase.
https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/tm2/pkg/std/std.proto#L18-L27

Do you think that when an indexer shows a transaction, it should conform to the indexer's convention? 
Or is there a possibility to unify the convention in the chain?

File: serve/graph/schema/filter/transaction_filter.graphql (hunk 14 above)
[the remark this answers is not in the document]
Comment (ajnavarro) [reply serve/graph/schema/filter/transaction_filter.graphql:89]: I see your point. With my approach won't be possible to get specific message types that are not sharing params. LGTM.

**Comment by ajnavarro** [comment]
I had been playing with the new GraphQL API and it is amazing the amount of info you can get, and how fast it is. I just added minor comments on schema parameters using uppercase for consistency.

:tada:

**Review by zivkovicmilos — APPROVED** [review]
I've gone through @ajnavarro's comments after my own review, and have to say that this looks amazing 💯 

Thank you for contributing back to the tool, it really means a lot to the team 🙏
